### ISSUE TITLE
Simple Stack's Future

### ISSUE THREAD

**Issue by Mikedeejay2** [Mikedeejay2/SimpleStackPlugin#83]
I'm creating this sticky issue to explain where Simple Stack is as a plugin, what's currently happening with it, and where I want it to be in the future.

### Simple Stack's current state
Simple Stack has been on version 1.3.x for over 6 months. Work has been put into version 1.4 but after a few thousand lines of code I'm just not happy with the way it has been implemented and I feel like it's not a safe future. Simple Stack may have "Simple" in the title but internally it is turning out to be very complex. It shouldn't be too difficult for a programmer or experienced player to realize that version 1.3.x is not stable. Version 1.3.x of Simple Stack is still fundamentally built on code that was thrown together so I could originally stack potions on a creative server. In version 1.4 I attempted to take a different approach of completely recoding Minecraft's item movement system from scratch. This worked somewhat better but the usage of cancelling events to move the items still causes issues such as villager trading not rewarding XP to the villager because the event is cancelled. Issues like the villager XP was fixed temporarily using NMS code but with the big changes of 1.17 and the global dislike of using NMS in Minecraft plugins I've scrapped that solution for the time being. Simple Stack 1.4 isn't as clean of a solution as I was imagining when I started coding it back in January, and that's why I'm most likely going to be scrapping that implementation. 

### Should I still use version 1.3.x of Simple Stack until a new version comes out?
This depends on the type of server you're intending on running Simple Stack on. Running Simple Stack on creative servers is fine, there's no reason to worry about players finding duplication glitches or running into villagers not being able to level up. For survival servers, whether or not you should run Simple Stack is heavily dependent on what type of server it is. A small survival server with only friends would be fine given that players don't want to level up villagers. On any public server that is survival/factions/anarchy/etc, I would strongly recommend not running Simple Stack because of the risks that a player could find a duplication glitch or lose items because of Simple Stack. 

**TLDR**: Private/testing/creative servers should be fine running Simple Stack, Public/survival/factions/anarchy/etc runs the risk of players experiencing bugs with item movement.

### What is currently happening with Simple Stack
Simple Stack will continue to receive bug fixes and version updates to version 1.3.x. Bugs will be fixed given that I have the ability to fix them with the current state of the source code. **If you have a bug that you'd like to report, please report it.** I will still read, assess, and comment on any bug reports that I receive. Given that I can fix the issue with the existing code I will fix the bug and release a new version of 1.3.x. Given that I can not fix the bug in 1.3.x I will ensure that the bug is fixed in the next major version.

Bugs that can not be fixed in 1.3:
* Villager related bugs (Villagers not levelling, villagers not replenishing trades, etc, accurately levelling villagers or notifying villagers of a trade can only be done through NMS as far as I'm aware)
* XP dropped by trading with villagers or using grindstones (This XP can only be accurately calculated through NMS code as far as I'm aware)
* Some compatibility issues (ChestSort not correctly sorting by SimpleStack item amounts, etc. I could theoretically fix this but the amount of time and code that would be required is too much)

### Where I want Simple Stack to be in the future
Simple Stack's item code will need to be rewritten. **With the amount of code required for a new system, I'm scrapping version 1.4.0 and going to version 2.0.**

Work hasn't started on a new version and I don't have everything planned on how it will work yet but I have a few things that I want to ensure:
1. Bukkit Events aren't cancelled. This causes a lot of compatibility issues and general design issues that could be avoided if the events weren't cancelled. Issues with villager levelling, blocks not granting XP, and much more are caused by the cancellation of events.
2. Simple Stack functions at a level below events. Bukkit events were meant for cancelling actions if need be, not replacing them entirely.
3. Possibly the ability to stack above 64? I haven't looked at how item amount values are stored and managed below the Bukkit API, but depending on Minecraft's implementation I might be able to push that limit to 127 if everything works out. Item amounts are serialized as a byte (max of 127) but are stored at runtime as an int (max of 2,147,483,647). If I can figure out how to make Minecraft serialize item amounts as an int instead of a byte the ability to store 2 billion items in a stack might just be reachable. I, however, can't modify anything from the client so if the item amount sent to the client is interpreted as a byte I can't do anything about it.
4. No direct use of NMS, as in not importing NMS as a dependency. Access through reflection, ASM, or other sources is most likely inevitable without the use of Bukkit events.

### What this will look like
Branch `master/dev` will be moved to branch `scrapped/1.4.0-update`, versions 1.3.x will change from branch `master/stable` will remain on the stable branch until new stable versions for 2.0. Branch `master/dev` will be created as the master branch which will be the newest code that I've worked on, tests, etc.

### ETA on a release?
None. I can't gauge how long something I haven't started on will take. It definitely won't be soon, coding it robustly is the most important thing right now.

**Comment by Mikedeejay2** [comment]
This issue is very outdated, closing for now. I may create an updated pinned issue at a later point
